docs/src/proposals/tiered-accounts-db-storage.md

```diff
@@ -0,0 +1,466 @@
+    +------------------+-----------------------------------------------+
+    | Avg Account Data |     270 bytes    |   270 bytes  |  108 bytes  |
+    +------------------------------------------------------------------+
+    | Avg Account Size |     410 bytes    |   342 bytes  |  184 bytes  |
+    +------------------------------------------------------------------+
+    | Relative Size    |       100%       |      83%     |      44%    |
+    +------------------------------------------------------------------+
+
+
+### The Detailed Design
+
+The proposed design uses a extensible blocked-based file format that
+can be used for both cold and hot storage.  The format includes
+five sections: Account Data Blocks, Account Metas Block, Owners Block,
+Account Index Block, and Footer.
+
+The Footer includes information about to read the file.  Specifically,
+it includes the file format version and the format of each section.
+This allows cold and hot storage shares the same file layout but optimize
+differently in each section.  For instance, cold storage has its account
+data blocks compressed while hot storage stores its account data blocks
+in uncompressed and aligned format in order to make it memory-mapped
+possible.
+
+    +------------------------------------------------------------------------------+
+    | Account Data Blocks      | Account data in compressed blocks.                |
+    | (compressed)             | Optional fields such as rent_epoch are also       |
+    |                          | stored here.                                      |
+    +------------------------------------------------------------------------------+
+    | Account Metas Block      | Fixed-size entries for storing accounts metadata. |
+    | (can be compressed)      | Metadata also includes information about how to   |
+    |                          | access its owner and account data.                |
+    +------------------------------------------------------------------------------+
+    | Owners Block             | 32 byte (Pubkey) per entry                        |
+    | (uncompressed)           | Each account meta entry has an index pointing     |
+    |                          | to its owner entry.                               |
+    +------------------------------------------------------------------------------+
+    | Account Index Block      | Index for accessing accounts meta block           |
+    | (uncompressed)           | Footer has an enum entry describing the format    |
+    |                          | of this block.                                    |
+    +------------------------------------------------------------------------------+
+    | Footer                   | Stores information about how to read the file.    |
+    | (uncompressed)           | (e.g. the offset and format of each block.)       |
+    +------------------------------------------------------------------------------+
```

Comment:
> In order to get the footer we always have to query the file size from the filesystem/OS, which means a syscall. And I'd prefer to not require a syscall that can otherwise be avoided.

This makes sense, but when opening an accounts file, we also want to make sure whether the file is truncated or not.  In this case, we still need some type of seek operation to the end in order to achieve that.

One thing to mention is that the seek to the end of the file will only be performed once when a file is newly opened. After that, the footer content will be cached.

Reply:
> I think one thing I'm still struggling with is that I'm not aware of any other file format that uses a footer instead of a header. I would think that if performance is demonstrably better with a footer then we'd see that being the default for file formats.

RocksDB's SST file format is one example.

```
<beginning_of_file>
[data block 1]
[data block 2]
...
[data block N]
[meta block 1: filter block]                  (see section: "filter" Meta Block)
[meta block 2: index block]
[meta block 3: compression dictionary block]  (see section: "compression dictionary" Meta Block)
[meta block 4: range deletion block]          (see section: "range deletion" Meta Block)
[meta block 5: stats block]                   (see section: "properties" Meta Block)
...
[meta block K: future extended block]  (we may add more meta blocks in the future)
[metaindex block]
[Footer]                               (fixed size; starts at file_size - sizeof(Footer))
<end_of_file>
```

https://github.com/facebook/rocksdb/wiki/Rocksdb-BlockBasedTable-Format

> Where is the content cached?

When a file is newly opened, the footer struct is a member field of the tiered storage AccountsFile.  Since footer contains all the offsets of each important section, it makes sense to make it same life-cycle as its AccountsFile so that it can quickly answer many questions without the need of loading additional block each time.

```
 pub struct HotStorageReader {
     map: Mmap,
     footer: TieredStorageFooter,
 }
```

Reply:
Offline discussed with @jeffwashington and @brooksprumo.  We will go with the footer design.